src/catalog/postgres/pg_statistic_impl.cpp

```diff
@@ -129,4 +130,137 @@ bool PgStatisticImpl::DeleteColumnStatistics(const common::ManagedPointer<transa
+}
+
+std::unique_ptr<optimizer::TableStats> PgStatisticImpl::GetTableStatistics(
+    common::ManagedPointer<transaction::TransactionContext> txn,
+    common::ManagedPointer<DatabaseCatalog> database_catalog, table_oid_t table_oid) {
+  const auto &oid_pri = statistic_oid_index_->GetProjectedRowInitializer();
+  const auto &oid_prm = statistic_oid_index_->GetKeyOidToOffsetMap();
+
+  byte *const buffer = common::AllocationUtil::AllocateAligned(pg_statistic_all_cols_pri_.ProjectedRowSize());
+  byte *const key_buffer = common::AllocationUtil::AllocateAligned(oid_pri.ProjectedRowSize());
+
+  // Look for the column statistic in pg_statistic_index.
+  std::vector<storage::TupleSlot> index_results;
+  {
+    auto *pr = oid_pri.InitializeRow(buffer);
+    auto *pr_hi = oid_pri.InitializeRow(key_buffer);
+
+    // Low key (class, min col_oid_t)
+    pr->Set<table_oid_t, false>(oid_prm.at(indexkeycol_oid_t(1)), table_oid, false);
+    pr->Set<col_oid_t, false>(oid_prm.at(indexkeycol_oid_t(2)), col_oid_t(std::numeric_limits<uint32_t>::min()), false);
+
+    // High key (class + 1, max col_oid_t)
+    pr_hi->Set<table_oid_t, false>(oid_prm.at(indexkeycol_oid_t(1)), table_oid + 1, false);
+    pr_hi->Set<col_oid_t, false>(oid_prm.at(indexkeycol_oid_t(2)), col_oid_t(std::numeric_limits<uint32_t>::max()),
+                                 false);
+
+    statistic_oid_index_->ScanAscending(*txn, storage::index::ScanType::Closed, 2, pr, pr_hi, 0, &index_results);
+  }
+  NOISEPAGE_ASSERT(!index_results.empty(), "Every table should have column stats");
+
+  std::vector<std::unique_ptr<optimizer::ColumnStatsBase>> col_stats_list;
+
+  auto all_cols_pr = common::ManagedPointer(pg_statistic_all_cols_pri_.InitializeRow(buffer));
+  for (const auto &slot : index_results) {
+    statistics_->Select(txn, slot, all_cols_pr.Get());
+
+    auto col_oid = *PgStatistic::STAATTNUM.Get(all_cols_pr, pg_statistic_all_cols_prm_);
+    auto type = database_catalog->GetSchema(txn, table_oid).GetColumn(col_oid).Type();
+    col_stats_list.emplace_back(CreateColumnStats(all_cols_pr, table_oid, col_oid, type));
+  }
+  delete[] buffer;
+  delete[] key_buffer;
+  return std::make_unique<optimizer::TableStats>(db_oid_, table_oid, &col_stats_list);
+}
+
+std::unique_ptr<optimizer::ColumnStatsBase> PgStatisticImpl::CreateColumnStats(
+    common::ManagedPointer<storage::ProjectedRow> all_cols_pr, table_oid_t table_oid, col_oid_t col_oid,
+    type::TypeId type) {
+  auto num_rows = *PgStatistic::STA_NUMROWS.Get(all_cols_pr, pg_statistic_all_cols_prm_);
+  auto non_null_rows = *PgStatistic::STA_NONNULLROWS.Get(all_cols_pr, pg_statistic_all_cols_prm_);
+  auto distinct_values = *PgStatistic::STA_DISTINCTROWS.Get(all_cols_pr, pg_statistic_all_cols_prm_);
+  const auto *top_k_str = PgStatistic::STA_TOPK.Get(all_cols_pr, pg_statistic_all_cols_prm_);
+  const auto *histogram_str = PgStatistic::STA_HISTOGRAM.Get(all_cols_pr, pg_statistic_all_cols_prm_);
+
+  switch (type) {
```

Comment:
Is there a better way to do this? I need to convert a `type::TypeId` to a `execution::sql::val`, but couldn't find a cleaner way other than hardcoding.